keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,223 @@
+project could mandate the use of those labels in a conformant distribution, since we anticipated that many deployments
+of Kubernetes would have more nuanced control-plane topologies than simply "a control plane node".
+
+Over time, several changes to Kubernetes core and related projects were introduced that depended on the
+`node-role.kubernetes.io/master` label to vary their behavior in contravention to the guidance the label was
+approved under. This was unintentional and due to unclear reviewer guidelines that have since been more strictly
+enforced. Likewise, the complexity of Kubernetes deployments has increased and the simplistic mapping of control
+plane concepts to a node has proven to limit the ability of conformant Kubernetes distributions to self-host, as
+anticipated.
+
+
+### Goals
+
+This KEP:
+
+* Clarifies that the use of the `node-role.kubernetes/*` label namespace is reserved solely for end-user and
+  external Kubernetes consumers, and:
+  * May not be used to vary behavior within Kubernetes projects that are subject to API review (kubernetes/kubernetes
+    and all components that expose APIs under the `*.k8s.io` namespace)
+  * Is not required to be used for a cluster to be conformant
+* Describes the locations within Kubernetes that must be changed to use an alternative mechanism for behavior
+  * Suggests approaches for each
+* Describes the timeframe and migration process for Kubernetes distributions and deployments to update labels
+
+
+## Proposal
+
+### Use of `node-role.kubernetes.io/*` labels
+
+* Kubernetes components MUST NOT set or alter behavior on any label within the `node-role.kubernetes.io/*` namespace.
+* Kubernetes components (such as `kubectl`) MAY simplify the display of `node-role.kubernetes.io/*` labels to convey the node roles of a node
+* Kubernetes examples and documentation MUST NOT leverage the node-role labels for node placement
+* External users, administrators, conformant Kubernetes distributions, and extensions MAY use `node-role.kubernetes.io/*` without reservation
+  * Extensions are recommended not to vary behavior based on node-role, but MAY do so as they wish
+* First party components like `kubeadm` MAY use node-roles to simplify their own deployment mechanisms.
+* Conformance tests MUST NOT depend on the node-role labels in any fashion
+
+
+### Migrating existing deployments
+
+The proposed fixes will all require deployment-level changes. That must be staged across several releases, and it should be possible for
+deployers to move early and "fix" the issues that may be caused by their topology.
+
+Therefore, for each change we recommend the following process to adopt the new labels in successive releases:
+
+* Release 1 (1.16):
+  * Introduce a feature gate for disabling node-role being honored. The gate defaults to on. `NodeRoleBehavior=on`
+  * Define the new node label with an associated feature gate for each feature area. The gate defaults to off. `FeatureA=off`
```

Comment:
for node-specific labels that modify non-security-sensitive behavior, should we choose labels within the `*.node.kubernetes.io/*` or `*.kubelet.kubernetes.io/*` namespaces to allow use with `--node-labels`?

Reply:
Yes.